### PR TITLE
chore(aws_sqs source): Convert decoder to iterator

### PR DESCRIPTION
**Pull request by bruceg**
The message decoder in the `aws_sqs` source does a lot of `async` work for no real gain, as none of the code paths will ever actually wait asynchronously. This PR converts it all to an iterator.

The bulk of the code additions are the `FoldFinally` iterator adapter, for which I couldn't find an equivalent. The stream handles it by emitting the final value after the last `yield` with an internalized state machine built up by the compiler. This is an equivalent mechanism for an iterator, exposed as a struct.

This is partially related to #11908, as this kind of work may be required in other places if we have to move away from tokio's `FramedRead`.

### CODE REVIEW

File: src/sources/aws_sqs/source.rs [hunk 1]
```diff
@@ -167,44 +165,106 @@ async fn delete_messages(client: &SqsClient, receipts: &[String], queue_url: &st
...
+        };
+    })
+    .flatten()
+}
+
+trait FoldFinallyExt: Sized {
+    /// This adapter applies the `folder` function to every element in
+    /// the iterator, much as `Iterator::fold` does. However, instead
+    /// of returning the resulting folded value, it calls the
+    /// `funally` function after the last element. This function
```
Comment (pablosichert) [inline src/sources/aws_sqs/source.rs:218]:
```suggestion
    /// `finally` function after the last element. This function
```

😄

or rather:

```suggestion
    /// `fold_finally` function after the last element. This function
```

Comment (spencergilbert) [inline src/sources/aws_sqs/source.rs:218]:
```suggestion
    /// `finally` function after the last element. This function
```

**Review by pablosichert — APPROVED** [review]
Very nice catch!

I think we have a couple more places where we use `FramedRead` by converting an iterator into a stream because we are in an `async` context anyway, but should probably use an iterator to reduce load on the Tokio scheduler like in this example.

File: src/sources/aws_sqs/source.rs (hunk 1 above)
Comment (bruceg) [reply src/sources/aws_sqs/source.rs:218]: 🤦🏻

**Pull request merged**